modin/data_management/functions/mapreducefunction.py

```diff
@@ -20,11 +20,20 @@ def call(cls, map_function, reduce_function, **call_kwds):
         def caller(query_compiler, *args, **kwargs):
             preserve_index = call_kwds.pop("preserve_index", True)
             axis = call_kwds.get("axis", kwargs.get("axis"))
+            if kwargs.get("numeric_only", "is_not_exist") == "is_not_exist":
```

Comment:
Why is this change necessary? `numeric_only` should be handled at a higher level than this.

Reply:
```python
import pandas
import numpy as np
# plain df
df = pandas.DataFrame(
    {
        "int_col": [-5, 2, 7, 16],
        "float_col": [np.NaN, -9.4, 10.1, np.NaN],
        "str_col": ["a", np.NaN, "c", "d"],
        "bool_col": [False, True, True, False],
    }
)
df
   int_col  float_col  bool_col
0       -5        NaN     False
1        2       -9.4      True
2        7       10.1      True
3       16        NaN     False
# it is that what is happening during our `map` phase
# pandas 1.1.5 behavior:
df.max(axis=0, numeric_only=True)
int_col      16.0
float_col    10.1
bool_col      1.0
dtype: float64
# pandas 1.2.1 behavior:
df.max(axis=0, numeric_only=True)
int_col        16
float_col    10.1
bool_col     True
dtype: object

# the difference is in last two lines
bool_col      1.0
dtype: float64
# and
bool_col     True
dtype: object
```
Since we perform `max` and `min` through 2 stages (`map` and `reduce`) we need to pass `not numeric_only` to `reduce` phase.
I am not sure that we can handle this at a higher level. Some changes are required here in any case.

Reply:
pandas issue [39607](https://github.com/pandas-dev/pandas/issues/39607).

Reply:
One more of the solutions might be passing parameter `_numeric_only` to `map_reduce_function` and doing similar logic that is now.

Reply:
If we need to convert types, thatThere can be done with `as_type`. There shouldn't be any addition to this file for changing the type implicitly in this way. It will also cause `dtypes` to be an issue, so it's best handled at the higher layers.

Reply:
`astype` can't be used at the API layer because we need to handle the case at the time point between `map` and `reduce` phases. Even if we could use `astype` it could add some overhead. So, moved the handling to the query compiler.

Reply:
If we convert the types beforehand, it will be correct in the output. There should be no need to handle anything in between map and reduce phases if the types are handled first.

Reply:
I am not sure when we can convert the types beforehand. If we convert them at the API level as `data.astype(object)`, then we have empty dataframe after map phase that is incorrect.

Reply:
And if we set `numeric_only=False/None`, then we have other failures.

Reply:
This will change only bool to float: `self.astype({c: "float" for c in self.columns if self.dtypes[c] == np.bool})`

Reply:
Okay, I see, but then we will have a behavior that match pandas 1.1.5 behavior mentioned above. Also, some overhead will be added because of use of `astype`. Are we okay with this?

Reply:
Okay, I see, but then we will have a behavior that matches pandas 1.1.5 behavior mentioned above. Also, some overhead will be added because of use of astype. Are we okay with this?

Reply:
What is the current Modin behavior in case of bool for `numeric_only=True` if the extra handling of `numeric_only` is removed from this?

Reply:
If we perform `self.astype({c: "float" for c in self.columns if self.dtypes[c] == np.bool})` and leave `max = MapReduceFunction.register(pandas.DataFrame.max)`, we will have the behavior matching pandas 1.1.5 for both `numeric_only=True` and `numeric_only=False`.

Reply:
We need behavior to match pandas 1.2, what if we do not do the `astype`? What behavior in Modin was changed here? Is the dtype of the result different depending on `numeric_only`?

Reply:
If we do not do the `astype` and leave the changes as it is in the PR, we match pandas 1.2.1 behavior. But if we do the `astype`, we match 1.1.5 behavior regardless the processing of `numeric_only` (True/False/None). So, I would leave the changes as they are now.

Reply:
I see, I had not seen that the changes to `MapReduceFunction` were reverted, so the current way of handling it is fine.